niri-config/src/lib.rs

```diff
@@ -3022,13 +3086,28 @@ where
 
         let mut binds = Vec::new();
 
+        let mod_key = ctx.get::<ModKey>().cloned();
+
         for child in node.children() {
             match Bind::decode_node(child, ctx) {
                 Err(e) => {
                     ctx.emit_error(e);
                 }
                 Ok(bind) => {
-                    if seen_keys.insert(bind.key) {
+                    let Some(mod_key) = &mod_key else {
+                        // On first config pass, we don't know what the ModKey is, so don't check
+                        // for bind collisions.
+                        binds.push(bind);
+                        continue;
+                    };
+
+                    let mut key = bind.key;
+                    if key.modifiers.contains(Modifiers::COMPOSITOR) {
+                        key.modifiers.remove(Modifiers::COMPOSITOR);
+                        key.modifiers.insert(mod_key.0);
+                    }
+
+                    if seen_keys.insert(key) {
```

Comment:
Why all this new logic? It works fine now (even with mod key being different between host and nested), so why do this with configurable mod key?

Reply:
While the current setup works, there's an edge-case bug I'd like to address. For example, with a config like this:

```
input {
  mod-key "super"
}

binds {
  Mod+Z { spawn "kitty"; }
  Super+Z { show-hotkey-overlay; }
}
```

The parser can't detect conflicts between Mod+Z and Super+Z unless it knows what `Mod` maps to.

Here's the approach I propose:

1. Parse the config once (without checking for conflicting binds).
2. After parsing, validate the binds with the the `mod-key` it just extracted from the config.
3. If duplicates exist, re-parse with `ModKey` passed via the context to generate a `DecodeError` (leveraging knuffel's error formatting).

This way, we avoid custom error handling and ensure conflicts are reported accurately by the parser.
What do you think?

Reply:
> and it shouldn't

Why not? In this scenario, `Mod+Z` and `Super+Z` clearly conflict, so I'd expect it to report an error.

> You don't want your config to suddenly break when you open nested for example.

You can use `NIRI_OVERRIDE_MOD` for developing/testing niri. Do you think that's good enough?

Reply:
> You can use NIRI_OVERRIDE_MOD for developing/testing niri. Do you think that's good enough?

See my comment in the other thread on why I don't think that's a good solution.

Reply:
What about the conflicts question? My current implementation allows multiple modifiers in `mod-key`:

```
input {
  mod-key "ctrl+shift"
}

binds {
  Mod+Z { spawn "kitty"; }
  // this conflicts with the bind above
  Ctrl+Shift+Z { show-hotkey-overlay; }
}
```

Should I simplify the implementation and only allow a single modifier for `mod-key`?

Reply:
I think you should allow only single if that's fine. And I think I would leave conflict resolution out, at least for this PR.

Reply:
Sounds good. I'll take another crack at it later.